internal/transport/http2_server.go

```diff
@@ -966,43 +966,52 @@ func (t *http2Server) keepalive() {
 				// The connection has been idle for a duration of keepalive.MaxConnectionIdle or more.
 				// Gracefully close the connection.
 				t.drain(http2.ErrCodeNo, []byte{})
-				// Resetting the timer so that the clean-up doesn't deadlock.
-				maxIdle.Reset(infinity)
 				return
 			}
-			maxIdle.Reset(val)
-		case <-maxAge.C:
+			idleTimer.Reset(val)
+		case <-ageTimer.C:
 			t.drain(http2.ErrCodeNo, []byte{})
-			maxAge.Reset(t.kp.MaxConnectionAgeGrace)
+			ageTimer.Reset(t.kp.MaxConnectionAgeGrace)
 			select {
-			case <-maxAge.C:
+			case <-ageTimer.C:
 				// Close the connection after grace period.
 				infof("transport: closing server transport due to maximum connection age.")
 				t.Close()
-				// Resetting the timer so that the clean-up doesn't deadlock.
-				maxAge.Reset(infinity)
 			case <-t.done:
 			}
 			return
-		case <-keepalive.C:
-			if atomic.CompareAndSwapUint32(&t.activity, 1, 0) {
-				pingSent = false
-				keepalive.Reset(t.kp.Time)
+		case <-kpTimer.C:
+			lastRead := atomic.LoadInt64(&t.lastRead)
+			if lastRead > prevNano {
+				// There has been read activity since the last time we were
+				// here. Setup the timer to fire at kp.Time seconds from
+				// lastRead time and continue.
+				outstandingPing = false
+				kpTimer.Reset(time.Duration(lastRead) + t.kp.Time - time.Duration(time.Now().UnixNano()))
+				prevNano = lastRead
 				continue
 			}
-			if pingSent {
+			if outstandingPing && kpTimeoutLeft <= 0 {
 				infof("transport: closing server transport due to idleness.")
 				t.Close()
-				// Resetting the timer so that the clean-up doesn't deadlock.
-				keepalive.Reset(infinity)
 				return
 			}
-			pingSent = true
-			if channelz.IsOn() {
-				atomic.AddInt64(&t.czData.kpCount, 1)
+			if !outstandingPing {
+				if channelz.IsOn() {
+					atomic.AddInt64(&t.czData.kpCount, 1)
+				}
+				t.controlBuf.put(p)
+				kpTimeoutLeft = t.kp.Timeout
+				outstandingPing = true
 			}
-			t.controlBuf.put(p)
-			keepalive.Reset(t.kp.Timeout)
+			// The amount of time to sleep here is the minimum of kp.Time and
+			// timeoutLeft. This will ensure that we wait only for kp.Time
+			// before sending out the next ping (for cases where the ping is
+			// acked).
+			sleepDuration := minTime(t.kp.Time, kpTimeoutLeft)
+			kpTimeoutLeft -= sleepDuration
+			prevNano = lastRead
```

Comment:
I this case I think we know `prevNano == lastRead` already (otherwise the first `if` above would have skipped this), so this is unnecessary.  Or `lastRead` is going backwards which is :open_mouth: .

Reply:
Benchmark config::

    Mode: Unary
    BenchTime: 1m
    MaxConcurrentCalls: 100
    ReqSize: 1B
    RespSize: 1B

Benchmark results::

               Title       Before        After Percentage
            TotalOps      2240700      2237780    -0.13%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8491.83      8492.49     0.01%
           Allocs/op       161.22       161.28     0.00%
             ReqT/op    298760.00    298370.67    -0.13%
            RespT/op    298760.00    298370.67    -0.13%
            50th-Lat   2.601787ms   2.619027ms     0.66%
            90th-Lat   3.766543ms   3.750905ms    -0.42%
            99th-Lat   5.282168ms   5.185314ms    -1.83%
             Avg-Lat   2.676142ms   2.679648ms     0.13%

Reply:
Why do you say `lastRead` is going backwards?
Looks like I added this assignment in the client side PR because we wanted to avoid a crazy race. But thinking about it now, it looks like we don't need to make the assignment. I'm removing it here and in the client side.

Reply:
> Why do you say lastRead is going backwards?

I didn't say it is.  I said the only way we can reach this point is _if_ `lastRead` is either unchanged or going backwards.  I assumed the latter wasn't possible, so `lastRead == prevNano` and there's no need for this assignment.